### PR TITLE
Autodetect cudnn-fp16 and re-enable --gpu option

### PR DESCRIPTION
**Pull request by borg323**
This pr does several things:
1. Checks that lc0 runs early, before downloading anything
2. If no --backend-opts are given, and cudnn is available, cudnn-fp16 is tried first and if no parallelism is given 32 is used with it.
3. If cudnn-fp16 is not available, the cudnn bckend is used immediately
4. the --gpu option can select gpu, but only if --backend-opts is not used.

### CODE REVIEW

File: lc0_main.go [hunk 1]
```diff
@@ -242,9 +271,16 @@ func (c *cmdWrapper) launch(networkPath string, args []string, input bool) {
 			}
 		}
 		c.Cmd.Args = append(c.Cmd.Args, fmt.Sprintf("--backend-opts=%s", *backopts))
+	} else if hasCudnnFp16 {
```
Comment (Tilps) [inline lc0_main.go]: Needs to have a hasCudnn path after this - otherwise gpu target will be ignored for cudnn.

Comment (borg323) [reply lc0_main.go]: If cudnn-fp16 fails the error handler hardcodes backend=cudnn in backopts. 
However it is a good idea and cleaner to just set hasCudnnFp16 to false in the error handler.

Comment (Tilps) [reply lc0_main.go]: I guess I was talking about a hypothetical build, built without fp16 support.

**Comment by borg323** [comment]
Unfortunately, lc0 throws an exception for unsupported fp16 in selfplay mode that isn't printing anything on stderr in windows. Without a way to handle this, the patch can't be accepted.

**Comment by borg323** [comment]
Added a second test for the exception (that works on windows and linux) and fixed a crash due to a stupid mistake. I think this is now ready, but ideally would like some macos testing as well.

**Pull request merged**